### PR TITLE
unicode functions

### PR DESCRIPTION
**Pull request by AviAvni**
fix https://github.com/RedisGraph/RedisGraph/issues/2670 https://github.com/RedisGraph/RedisGraph/issues/2796

### CODE REVIEW

File: src/util/strutil.c [hunk 1]
```diff
@@ -8,6 +8,19 @@
...
 #include "RG.h"
 #include "rmalloc.h"
+#include "deps/utf8proc/utf8proc.h"
+
+int str_length(const char *str) {
+	size_t str_len = strlen(str);
+	utf8proc_int32_t c;
+	size_t i = 0;
+	int len = 0;
+	while(i < str_len) {
```
Comment (swilly22) [inline src/util/strutil.c:18]: I think we can do the following:
`str_len` is the length in bytes (should be rename to reflect that!)
`i = utf8proc_iterate(str + i, -1, &c);` gives us the number of bytes for a single character
`i / str_len` should be the length of the string in characters

File: src/util/strutil.c [hunk 2]
```diff
@@ -17,8 +30,14 @@ void str_tolower(const char *str, char *lower, size_t *lower_len) {
...
 
+	utf8proc_int32_t c;
 	size_t i = 0;
-	for(; i < str_len; i++) lower[i] = tolower(str[i]);
+	int w = 0;
+	while(i < str_len) {
+		w = utf8proc_iterate(str + i, -1, &c);
+		utf8proc_encode_char(utf8proc_tolower(c), lower + i);
+		i+=w;
+	}
```
Comment (swilly22) [inline src/util/strutil.c:40]: can't we encode the entire string in one go (single function call) and then follow up with a single call to `to lower`?
I just hate to see this loop here...

File: src/util/strutil.c [hunk 3]
```diff
@@ -30,8 +49,14 @@ void str_toupper(const char *str, char *upper, size_t *upper_len) {
...
 
+	utf8proc_int32_t c;
 	size_t i = 0;
-	for(; i < str_len; i++) upper[i] = toupper(str[i]);
+	int w = 0;
+	while(i < str_len) {
+		w = utf8proc_iterate(str + i, -1, &c);
+		utf8proc_encode_char(utf8proc_toupper(c), upper + i);
+		i+=w;
+	}
```
Comment (swilly22) [inline src/util/strutil.c:59]: Similar comments as in `str_tolower`

File: tests/flow/test_function_calls.py [hunk 4]
```diff
@@ -1853,7 +1853,9 @@ def test65_ToLower(self):
         query_to_expected_result = {
             "RETURN toLower('MuChAcHo')": [['muchacho']],
             "RETURN toLower('mUcHaChO')": [['muchacho']],
-            "RETURN toLower(NULL)": [[None]]
+            "RETURN toLower(NULL)": [[None]],
+            "RETURN toLower('ΑαΒβΓγΔδΕεΖζΗηΘθΙιΚκΛλΜμΝνΞξΟοΠπΡρΣσςΤτΥυΦφΧχΨψΩω')": [["ααββγγδδεεζζηηθθιικκλλμμννξξοοππρρσσςττυυφφχχψψωω"]],
```
Comment (swilly22) [inline tests/flow/test_function_calls.py:1857]: Add comment explaining this is to test none ASCII characters

File: tests/flow/test_function_calls.py [hunk 5]
```diff
@@ -1862,7 +1864,9 @@ def test66_ToUpper(self):
         query_to_expected_result = {
             "RETURN toUpper('MuChAcHo')": [['MUCHACHO']],
             "RETURN toUpper('mUcHaChO')": [['MUCHACHO']],
-            "RETURN toUpper(NULL)": [[None]]
+            "RETURN toUpper(NULL)": [[None]],
+            "RETURN toUpper('ΑαΒβΓγΔδΕεΖζΗηΘθΙιΚκΛλΜμΝνΞξΟοΠπΡρΣσςΤτΥυΦφΧχΨψΩω')": [["ΑΑΒΒΓΓΔΔΕΕΖΖΗΗΘΘΙΙΚΚΛΛΜΜΝΝΞΞΟΟΠΠΡΡΣΣΣΤΤΥΥΦΦΧΧΨΨΩΩ"]],
+            "RETURN toUpper('АаБбВвГгДдЕеЖжЗзИиЙйКкЛлМмНнОоПпРрСсТтУуФфХхЦцЧчШшЩщЬьЭэЮюЯя')":  [["ААББВВГГДДЕЕЖЖЗЗИИЙЙККЛЛММННООППРРССТТУУФФХХЦЦЧЧШШЩЩЬЬЭЭЮЮЯЯ"]]
```
Comment (swilly22) [inline tests/flow/test_function_calls.py:1869]: Add comment explaining this is to test none ASCII characters

**Review by swilly22 — CHANGES_REQUESTED** [review]
I hope this is the only library we will need to handle none ASCII characters
Can we test the overhead of computing string length with this change?
Making sure we'll be able to use this library with the rest of the requirements on our roadmap.
Thank you!

**Comment by LiorKogan** [comment]
> I hope this is the only library we will need to handle none ASCII characters
I don't think that this library handles RegExs, and the library that handles RegExs may use a different UTF8 library...

Also, we'll need a function IsAlphaNumerric(utf8) for string.clean(...)

File: src/util/strutil.c [hunk 6]
```diff
@@ -8,6 +8,19 @@
 #include <ctype.h>
 #include "RG.h"
 #include "rmalloc.h"
+#include "deps/utf8proc/utf8proc.h"
```
[the remark this answers is not in the document]
Comment (rafie) [reply src/util/strutil.c:11]: We can just use `#include "utf8proc/utf8proc.h`
It will work ok.

File: src/util/strutil.c (hunk 1 above)
Comment (AviAvni) [reply src/util/strutil.c:18]: each character can be different size

File: src/util/strutil.c (hunk 2 above)
Comment (AviAvni) [reply src/util/strutil.c:40]: we can't

File: src/util/strutil.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (AviAvni) [reply src/util/strutil.c]: will implement all others functions in this PR

File: tests/flow/test_function_calls.py [hunk 7]
```diff
@@ -2104,7 +2085,9 @@ def test83_Replace(self):
             "RETURN replace('abcabc', 'bc', '0')": [["a0a0"]],
             "RETURN replace('abcabc', 'abc', '')": [[""]],
             "RETURN replace('abcabc', 'ab', '')": [["cc"]],
-            "RETURN replace('abcabc', '', '0')": [["0a0b0c0a0b0c0"]]
+            "RETURN replace('abcabc', '', '0')": [["0a0b0c0a0b0c0"]],
+            # test unicode charecters
+            "RETURN replace('丁丂七丄丅丆万丈三上', '\xe4', 'X')": [["丁丂七丄丅丆万丈三上"]],
```
Comment (swilly22) [inline tests/flow/test_function_calls.py:2090]: we're not expecting to see a change in the string, please add a comment explaining this and why.

File: src/util/strutil.h [hunk 8]
```diff
@@ -7,19 +7,41 @@
...
+	const char *str,
+	char *lower,
+	size_t *lower_len
+);
 
+// return utf8 string converted to lower case
+char *str_tolower
+(
+	const char *str
+);
```
Comment (swilly22) [inline src/util/strutil.h:33]: I'll assume passing an ASCII string to this function will also work, and the returned string will remain ASCII?
if that's the case please update the comment accordingly as it is a bit confusing in such case.

File: src/util/strutil.h [hunk 9]
```diff
@@ -7,19 +7,41 @@
...
 
-// Utility function to increase the size of a buffer.
-void str_ExtendBuffer(
+// return utf8 string converted to upper case
+char *str_toupper
+(
+	const char *str
+);
```
Comment (swilly22) [inline src/util/strutil.h:39]: same comment as for `str_tolower`

File: src/util/strutil.h [hunk 10]
```diff
@@ -7,19 +7,41 @@
...
+
+// utility function to increase the size of a buffer
+void str_ExtendBuffer
+(
 	char **buf,           // buffer to populate
 	size_t *bufferLen,    // size of buffer
 	size_t extensionLen   // number of bytes to add
 );
```
Comment (swilly22) [inline src/util/strutil.h:47]: Why do we need this function, who uses it?
isn't this a simple call to `realloc` ?

File: src/util/strutil.c [hunk 11]
```diff
@@ -8,65 +8,132 @@
 #include <ctype.h>
 #include "RG.h"
 #include "rmalloc.h"
-#include "deps/utf8proc/utf8proc.h"
+#include "utf8proc/utf8proc.h"
```
Comment (swilly22) [inline src/util/strutil.c:11]: I'll assume the folder `deps` is on the include path e.g. compiler flag `-I` ?

File: src/util/strutil.c [hunk 12]
```diff
@@ -8,65 +8,132 @@
...
 	*lower_len = str_len;
 
+	for(size_t i = 0; i < str_len; i++) lower[i] = tolower(str[i]);
+	lower[str_len] = 0;
+}
+
+bool str_utf8_validate
+(
+    const char *str
+) {
```
Comment (swilly22) [inline src/util/strutil.c:33]: missing documentation
if this is a "local" function please add `static`

File: src/util/strutil.c [hunk 13]
```diff
@@ -8,65 +8,132 @@
...
+
+    // while we didn't get to the end of the string
+    while(str[0] != 0) {
+        // increment current position by number of bytes in Unicode character
+        str += utf8proc_iterate((const utf8proc_uint8_t *)str, -1, &c);
+		if(c == -1) return false;
+    }
+
+    // return the length of the string in terms of Unicode characters
+    return true;
```
Comment (swilly22) [inline src/util/strutil.c:45]: misleading comment, function doesn't return the length of the string.

File: src/util/strutil.c [hunk 14]
```diff
@@ -8,65 +8,132 @@
...
+
+// determine the length of the string
+// not in terms of the number of bytes in the string
+// but in terms of the number of Unicode characters in the string
+int str_length
+(
+    const char *str
+) {
+	// hold current Unicode character
+    utf8proc_int32_t c;
```
Comment (swilly22) [inline src/util/strutil.c:56]: comment and variable do not seems to be aligned.

File: src/util/strutil.c [hunk 15]
```diff
@@ -8,65 +8,132 @@
...
+    return true;
+}
+
+// determine the length of the string
+// not in terms of the number of bytes in the string
+// but in terms of the number of Unicode characters in the string
+int str_length
+(
+    const char *str
+) {
```
Comment (swilly22) [inline src/util/strutil.c:54]: Will this function work on ASCII strings ?

File: src/util/strutil.c [hunk 16]
```diff
@@ -8,65 +8,132 @@
...
+	// hold current Unicode character
 	utf8proc_int32_t c;
-	size_t i = 0;
-	int w = 0;
-	while(i < str_len) {
-		w = utf8proc_iterate(str + i, -1, &c);
-		utf8proc_encode_char(utf8proc_tolower(c), lower + i);
-		i+=w;
+	const char *str_i = str;
```
Comment (swilly22) [inline src/util/strutil.c:78]: can we declare `str_i` as `const utf8proc_uint8_t *` ?

File: src/util/strutil.c [hunk 17]
```diff
@@ -8,65 +8,132 @@
...
+	size_t lower_len = 0;
+	char buffer[4];
+	while(str_i[0] != 0) {
+		str_i     += utf8proc_iterate((const utf8proc_uint8_t *)str_i, -1, &c);
+		lower_len += utf8proc_encode_char(utf8proc_tolower(c), (utf8proc_uint8_t *)buffer);
```
Comment (swilly22) [inline src/util/strutil.c:83]: are we expecting an upper case unicode character and its corresponding lower case counter part to differ in size?

File: src/util/strutil.c [hunk 18]
```diff
@@ -8,65 +8,132 @@
...
+        str += utf8proc_iterate((const utf8proc_uint8_t *)str, -1, &c);
+        // increment length of the string in terms of Unicode characters
+        len++;
+    }
+
+    // return the length of the string in terms of Unicode characters
+    return len;
+}
+
+char *str_tolower
```
Comment (swilly22) [inline src/util/strutil.c:72]: Will this function work for ASCII strings ?

File: src/arithmetic/string_funcs/string_funcs.c [hunk 19]
```diff
@@ -10,9 +10,10 @@
 #include "../../util/arr.h"
 #include "../../util/rmalloc.h"
 #include "../../util/uuid.h"
+#include "utf8proc/utf8proc.h"
 #include "../../util/strutil.h"
-#include "../../util/json_encoder.h"
 #include "../../datatypes/array.h"
+#include "../../util/json_encoder.h"
```
Comment (swilly22) [inline src/arithmetic/string_funcs/string_funcs.c:16]: as long as we're here, can you please order lines by their length ?

File: src/arithmetic/string_funcs/string_funcs.c [hunk 20]
```diff
@@ -35,9 +36,18 @@ SIValue AR_LEFT(SIValue *argv, int argc, void *private_data) {
 		// No need to truncate this string based on the requested length
 		return SI_DuplicateStringVal(argv[0].stringval);
```
Comment (swilly22) [inline src/arithmetic/string_funcs/string_funcs.c:37]: `SI_DuplicateStringVal` handles none ascii chars correctly ?

File: src/arithmetic/string_funcs/string_funcs.c [hunk 21]
```diff
@@ -35,9 +36,18 @@ SIValue AR_LEFT(SIValue *argv, int argc, void *private_data) {
...
-	char *left_str = rm_malloc((newlen + 1) * sizeof(char));
-	strncpy(left_str, argv[0].stringval, newlen * sizeof(char));
-	left_str[newlen] = '\0';
+
+	utf8proc_int32_t c;
+	int64_t newlen_bytes = 0;
+	const char *str = argv[0].stringval;
+	for (int i = 0; i < newlen; i++) {
+		newlen_bytes += utf8proc_iterate((const utf8proc_uint8_t *)(str+newlen_bytes), -1, &c);
+	}
```
Comment (swilly22) [inline src/arithmetic/string_funcs/string_funcs.c:45]: Please add a comment e.g.
`// determine new string byte size`

File: src/arithmetic/string_funcs/string_funcs.c [hunk 22]
```diff
@@ -160,9 +182,20 @@ SIValue AR_SUBSTRING(SIValue *argv, int argc, void *private_data) {
...
 
-	char *substring = rm_malloc((length + 1) * sizeof(char));
-	strncpy(substring, original + start, length);
-	substring[length] = '\0';
+	utf8proc_int32_t c;
+	int64_t start_bytes = 0;
+	for (int i = 0; i < start; i++) {
+		start_bytes += utf8proc_iterate((const utf8proc_uint8_t *)(original+start_bytes), -1, &c);
+	}
+
```
Comment (swilly22) [inline src/arithmetic/string_funcs/string_funcs.c:190]: consider `original += start_byte`
avoiding `original+start_bytes` within the following loop

File: src/arithmetic/string_funcs/string_funcs.c [hunk 23]
```diff
@@ -160,9 +182,20 @@ SIValue AR_SUBSTRING(SIValue *argv, int argc, void *private_data) {
 		length = MIN(length, suffix_len);
 	}
 
-	char *substring = rm_malloc((length + 1) * sizeof(char));
-	strncpy(substring, original + start, length);
-	substring[length] = '\0';
+	utf8proc_int32_t c;
+	int64_t start_bytes = 0;
+	for (int i = 0; i < start; i++) {
```
Comment (swilly22) [inline src/arithmetic/string_funcs/string_funcs.c:187]: for each one of the newly added loops, please add a comment explaining the purpose of the loop.

File: src/arithmetic/string_funcs/string_funcs.c [hunk 24]
```diff
@@ -296,6 +325,18 @@ SIValue AR_REPLACE(SIValue *argv, int argc, void *private_data) {
 	const char *ptr  = str;
 	const char **arr = array_new(const char *, 0);
 
+	if(!str_utf8_validate(old_string)) {
+		return SI_DuplicateStringVal(str);
```
Comment (swilly22) [inline src/arithmetic/string_funcs/string_funcs.c:329]: I don't understand what's going on here

File: src/arithmetic/string_funcs/string_funcs.c [hunk 25]
```diff
@@ -381,25 +424,33 @@ SIValue AR_SPLIT(SIValue *argv, int argc, void *private_data) {
...
-		
-		char *token  = strtok(str, delimiter);
-
-		if(!token) {
-			SIArray_Append(&tokens, argv[0]);
-			rm_free(str);
-			return tokens;
+		size_t rest_len   = str_len;
+		const char *start = str;
+		while(rest_len > delimiter_len) {
```
Comment (swilly22) [inline src/arithmetic/string_funcs/string_funcs.c:432]: Please add a comment explaining what this loop do

**Review by swilly22 — CHANGES_REQUESTED** [review]
Please see comment,
mostly requests for documentation

File: src/util/strutil.c (hunk 11 above)
Comment (AviAvni) [reply src/util/strutil.c:11]: yes

File: src/util/strutil.h (hunk 8 above)
Comment (LiorKogan) [reply src/util/strutil.h:33]: ASCII is a subset of UTF-8, not an alternative

File: src/util/strutil.c (hunk 17 above)
Comment (AviAvni) [reply src/util/strutil.c:83]: yes

File: src/arithmetic/string_funcs/string_funcs.c (hunk 20 above)
Comment (AviAvni) [reply src/arithmetic/string_funcs/string_funcs.c:37]: what is the diff?

**Review by swilly22 — APPROVED** [review]
